### DISCUSSION TITLE
Custom plugin - instances, concurrency and reloading

### DISCUSSION THREAD

**Discussion by luukes** [coredns/coredns#5868] [Q&A] [answered]
I'm writing a custom plugin that does request filtering and loads a few hundred MB of data. The plugin works fine, but there are open questions where I kindly ask for your guidance:

**Plugin Instances**
Because the loaded data is quite big I'm concerned that coredns will create **more than one** instance of my plugin. For what I can see in the code and logs so far, it's created only once, but I'm not sure if this could happen under load?

**Concurrency**
I assume that coredns starts many goroutines and accesses a plugins `ServeDNS()` concurrently. If this is correct, I need to make my plugin code thread safe, right?

**Reloading**
CoreDNS can be easily reloaded, but how can I catch the reload event in my plugin? Is there a default way to do this?

**Comment by chrisohaver** [comment, marked as answer]
These questions slipped by me until now. 

>  I'm concerned that coredns will create more than one instance of my plugin

Recent versions of coredns don't do this (older ones would but this was fixed in the 1.9.x's IIRC ).  High load would not affect this.

If someone is trying to respond to high load by scaling horizontally, they would do so by spinning up more coredns instances.  But this is an administrator action, not coredns doing it.

Another exception is during a reload, where momentarily the old plugin chain and new plugin chain are in memory at the same time, 
 
> I need to make my plugin code thread safe, right?

Yes.

> how can I catch the reload event in my plugin?

reload works by instantiating a new server with new instances of the plugins, and replacing the previous server. Usually there isn't a need for a plugin to catch this, since it would be destroyed and re-placed with a new instance.

**Reply by luukes** [reply]
Thanks Chris, this gives me more confidence.
So far the plugin works well, and as you've said, I don't need to handle reloading as a full CoreDNS reload also re-creates my plugin.
